_drafts/debug-monitor.md

```diff
@@ -0,0 +1,685 @@
+}
+```
+
+In one terminal you will need to start a gdbserver:
+
+```bash
+$ JLinkGDBServer  -if swd -device nRF52840_xxAA  -nogui
+```
+
+Then you need to compile the example application and flash it
+
+```bash
+$ cd ${INTERRUPT_REPO}/example/debugmon/
+$ make
+Compiling src/debug_monitor_exception.c
+[...]
+Linking library
+Generated build/nrf52.elf
+$ arm-none-eabi-gdb-py --eval-command="target remote localhost:2331"  --ex="mon reset" --ex="load"
+--ex="mon reset"  --se=build/nrf52.elf
+(gdb)
+```
+
+From GDB. we will disable halting debug so the program will continue running:
+
+```
+(gdb) set *(uint32_t*)0xE000EDF0=(0xA05F<<16)
+```
+
+In a final terminal we will open a serial connection using miniterm:
+
+```bash
+$ miniterm.py - 115200 --raw
+--- Available ports:
+...
+---  4: /dev/cu.usbmodem0006838593731 'J-Link'
+--- Enter port index or full name: 4
+--- Miniterm on /dev/cu.usbmodem0006838593731  115200,8,N,1 ---
+--- Quit: Ctrl+] | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H ---
+shell> debug_mon_en
+Monitor Mode Debug Enabled!
+shell> bkpt
+DebugMonitor Exception
+DEMCR: 0x01010000
+DFSR:  0x00000002 (bkpt=1, halt=0, dwt=0)
+Register Dump
+ r0  =0x00000001
+ r1  =0x20003310
+ r2  =0x73734900
+ r3  =0x000009d5
+ r12 =0xffffffff
+ lr  =0x0000094b
+ pc  =0x000009d4
+ xpsr=0x210e0000
+ ...
+```
+
+As expected we see bit 1 in `DFSR` is set indicating a breakpoint debug event was generated.
+We can use GDB to inspect the instruction located at`0x000009d4` and confirm it is indeed a breakpoint
+instruction:
+
+```
+(gdb) x/i 0x000009d4
+  0x9d4 <prv_issue_breakpoint>:	bkpt	0x0001
+```
+
+We can also dump the instruction and confirm it matches the BKPT instruction encoding (`0xbe..`):
+
+![]({% img_url breakpoint/bkpt-instruction-cortex-m.png %})
+
+```
+(gdb) p/x *(uint16_t*)0x000009d4
+$4 = 0xbe01
+```
+
+### Programmatically installing Breakpoints with the FPB
+
+In the FPB we can program one of the`FP_COMP` register to hold the address we would like to break
+on.
+
+The layout of the `FP_COMP` register is a little odd but basically for a given `instr_addr`, `FP_COMP[31:30]` map
+`instr_addr[1:0]`, `FP_COMP[28:2]` map to `instr_addr[28:2]`, and the lowest bit `FP_COMP[0]` controls
+whether the breakpoint is enabled or not.
+
+For the NRF52, we can transcribe the logic into code as follows:
+
+```c
+bool fpb_set_breakpoint(size_t comp_id, uint32_t instr_addr) {
+  if (instr_addr >= 0x20000000) {
+    // for revision 1 only breakpoints in code can be installed :/
+    return false;
+  }
+  // make sure the FPB is enabled
+  FPB->FP_CTRL |= 0x3;
+
+  const uint32_t replace = (instr_addr & 0x2) == 0 ? 1 : 2;
+  const uint32_t fp_comp = (instr_addr & ~0x3) | 0x1 | (replace << 30);
+  FPB->FP_COMP[comp_id] = fp_comp;
+  return true;
+}
+```
+
+> For more details on how to configure the Flash Patch And Breakpoint (**FPB**) unit,
+> [check out our post on the topic]({% post_url 2020-06-17-cortex-m-breakpoints %}#flash-patch--breakpoint-unit).
```

Comment:
I'd bring the name of the post in here rather than just "check out our post on the topic"